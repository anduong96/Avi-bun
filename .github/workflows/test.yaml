name: Tests

on:
  push:
    branches: [main]
  pull_request:
    branches: [main]

permissions:
  checks: write
  contents: write

jobs:
  run-tests:
    name: Run Tests
    runs-on: ubuntu-latest
    env:
      DOPPLER_TOKEN: ${{ secrets.DOPPLER_TOKEN }}
      ENV: test

    steps:
      - name: Check out Git repository
        uses: actions/checkout@v3

      - uses: oven-sh/setup-bun@v1
        with:
          bun-version: latest

      - name: 'Install dependencies'
        run: bun install --silent

      - name: Generate prisma schema
        run: bun run prisma:generate

      - name: Run tests
        run: bun run test:coverage
