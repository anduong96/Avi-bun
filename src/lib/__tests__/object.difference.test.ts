import { format } from 'sys';
import { ChangeType } from '@prisma/client';
import { describe, expect, test } from 'bun:test';

import { getObjectDifference } from '../objects/object.difference/get.object.difference';

describe('lib::object.difference', () => {
  test('should return an empty object when the current and previous objects are the same', () => {
    const current = { a: 1, b: 2 };
    const previous = { a: 1, b: 2 };
    const result = getObjectDifference(current, previous);
    expect(result).toMatchSnapshot();
  });

  test('should return an object with the keys and values that have changed between the current and previous objects', () => {
    const current = { a: 1, b: 2 };
    const previous = { a: 1, b: 3 };
    const result = getObjectDifference(current, previous);
    expect(result).toMatchSnapshot();
  });

  test('should return an object with the keys and values that have changed between the current and previous objects, using a key transform function', () => {
    const current = { a: 1, b: 2 };
    const previous = { a: 1, b: 3 };
    const result = getObjectDifference(current, previous);
    expect(result).toMatchSnapshot();
  });

  test('should return an object with the keys and values that have been removed between the current and previous objects', () => {
    type Entry = { a: number; b?: number; c?: number; d?: number };
    const current: Entry = { a: 1, b: 2, d: undefined };
    const previous: Entry = { a: 1, c: 3, d: 4 };
    const result = getObjectDifference(
      current,
      previous,
      (key, value, changeType) => {
        return changeType === ChangeType.REMOVED
          ? format('%s was removed', key.toUpperCase())
          : changeType === ChangeType.ADDED
          ? format('%s was added as %s', key.toUpperCase(), value)
          : changeType === ChangeType.MODIFIED
          ? format('%s was changed to %s', key.toUpperCase(), value)
          : undefined;
      },
    );

    expect(result).toMatchSnapshot();
  });
});
