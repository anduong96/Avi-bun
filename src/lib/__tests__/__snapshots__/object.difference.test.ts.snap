// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`lib::object.difference should return an empty object when the current and previous objects are the same 1`] = `[]`;

exports[`lib::object.difference should return an object with the keys and values that have changed between the current and previous objects 1`] = `
[
  {
    "changeType": "MODIFIED",
    "currentValue": 2,
    "description": "b was changed to 2",
    "originalValue": 3,
    "valueType": "NUMBER",
  },
]
`;

exports[`lib::object.difference should return an object with the keys and values that have changed between the current and previous objects, using a key transform function 1`] = `
[
  {
    "changeType": "MODIFIED",
    "currentValue": 2,
    "description": "b was changed to 2",
    "originalValue": 3,
    "valueType": "NUMBER",
  },
]
`;

exports[`lib::object.difference should return an object with the keys and values that have been removed between the current and previous objects 1`] = `
[
  {
    "changeType": "ADDED",
    "current": 2,
    "description": "B was added as 2",
    "previous": null,
    "valueType": "NUMBER",
  },
  {
    "changeType": "REMOVED",
    "currentValue": undefined,
    "description": "D was removed",
    "originalValue": 4,
    "valueType": "NUMBER",
  },
]
`;
